### ISSUE TITLE
need some sort of job epilog for wreck jobs

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#1497]
@trws mentioned a near term need for a kind of job epilog for wreck jobs. The specific use case is apparently for SCR to "drain" data off of local storage while also preventing the job from reaching "complete" state until all the data has been moved.

We could hook into the `rexecd_exit` handler for Lua scripts already exported by `wrexecd`, however I noticed that the job will be set to `complete` *before* these hooks are run.

```C
    if (exec_rc == 0) {
        rexec_state_change (ctx, "complete");
        wlog_msg (ctx, "job complete. exiting...");

        lua_stack_call (ctx->lua_stack, "rexecd_exit");
    }
```

The execution of all scripts `rexecd_exit` could be moved above `rexec_state_change` (hopefully with no side effects), or we could just add a call to a well known "epilog" script path, if it exists, before the job is "complete". Finally,  another idea would be to mirror slurm and set the job state to "completing" then run all exit handlers, then transition to "complete", but I believe that may cause cascading work in other subsystems/flux-sched, and we probably want to minimize the time spent on new features related to the old execution system.

**Comment by trws** [comment]
I could see it being useful to do it either way, but had been thinking in terms of running the epilogue before complete gets set. It may not really matter, but it seems less likely to cause users surprise and or noise to start there.


---
Sent from VMware Boxer

On May 1, 2018 at 3:20:04 PM PDT, Dong H. Ahn <notifications@github.com> wrote:

The specific use case is apparently for SCR to "drain" data off of local storage while also preventing the job from reaching "complete" state until all the data has been moved.

Is the reason we do this before the complete state so that another job won't be scheduled on those nodes while draining is happening?

I just talked to Adam Moody and he can also see the benefit of an alternative option, marking the computation job complete as soon as it finishes, that is, while the SCR's epilogue script is draining out the data may increase the job throughput. This was we can allow computation and data stage-out to be overlapped.

Finally, another idea would be to mirror slurm and set the job state to "completing" then run all exit handlers, then transition to "complete", but I believe that may cause cascading work in other subsystems/flux-sched, and we probably want to minimize the time spent on new features related to the old execution system.

Adjusting sched for this change should be trivial. If this makes your life easier, I won't object to do it this way.
